internal/mode/static/nginx/config/servers.go

```diff
@@ -248,41 +256,59 @@ func createLocations(server *dataplane.VirtualServer, serverID int) ([]http.Loca
 
-		for matchRuleIdx, r := range rule.MatchRules {
-			buildLocations := extLocations
-
-			if len(rule.MatchRules) != 1 || !isPathOnlyMatch(r.Match) {
-				intLocation, match := initializeInternalLocation(pathRuleIdx, matchRuleIdx, r.Match)
-				buildLocations = []http.Location{intLocation}
-				matches = append(matches, match)
+		if !needsInternalLocations(rule) {
+			for _, r := range rule.MatchRules {
+				extLocations = updateLocationsForFilters(r.Filters, extLocations, r, server.Port, rule.Path, rule.GRPC)
 			}
 
-			includes := createIncludes(r.Additions)
+			for i := range extLocations {
+				extLocations[i].Includes = createIncludesFromPolicyGenerateResult(
+					generator.GenerateForLocation(rule.Policies, extLocations[i]),
+				)
+			}
 
-			// buildLocations will either contain the extLocations OR the intLocation.
-			// If it contains the intLocation, the extLocations will be added to the final locations after we loop
-			// through all the MatchRules.
-			// It is safe to modify the buildLocations here by adding includes and filters.
-			buildLocations = updateLocationsForIncludes(buildLocations, includes)
-			buildLocations = updateLocationsForFilters(r.Filters, buildLocations, r, server.Port, rule.Path, rule.GRPC)
-			locs = append(locs, buildLocations...)
+			locs = append(locs, extLocations...)
+			continue
 		}
 
-		if len(matches) > 0 {
-			for i := range extLocations {
-				// FIXME(sberman): De-dupe matches and associated locations
-				// so we don't need nginx/njs to perform unnecessary matching.
-				// https://github.com/nginxinc/nginx-gateway-fabric/issues/662
-				var key string
-				if server.SSL != nil {
-					key = "SSL"
-				}
-				key += strconv.Itoa(serverID) + "_" + strconv.Itoa(pathRuleIdx)
-				extLocations[i].HTTPMatchKey = key
-				matchPairs[extLocations[i].HTTPMatchKey] = matches
+		internalLocations := make([]http.Location, 0, len(rule.MatchRules))
+
+		for matchRuleIdx, r := range rule.MatchRules {
+			if len(rule.MatchRules) != 1 || !isPathOnlyMatch(r.Match) {
+				intLocation, match := initializeInternalLocation(pathRuleIdx, matchRuleIdx, r.Match, grpc)
+				intLocation.Includes = createIncludesFromPolicyGenerateResult(
+					generator.GenerateForInternalLocation(rule.Policies),
+				)
+
+				intLocation = updateLocationForFilters(
+					r.Filters,
+					intLocation,
+					r,
+					server.Port,
+					rule.Path,
+					rule.GRPC,
+				)
+
+				internalLocations = append(internalLocations, intLocation)
+				matches = append(matches, match)
 			}
-			locs = append(locs, extLocations...)
 		}
+
+		httpMatchKey := serverID + "_" + strconv.Itoa(pathRuleIdx)
+		for i := range extLocations {
+			// FIXME(sberman): De-dupe matches and associated locations
+			// so we don't need nginx/njs to perform unnecessary matching.
+			// https://github.com/nginxinc/nginx-gateway-fabric/issues/662
+			extLocations[i].HTTPMatchKey = httpMatchKey
+			extLocations[i].Includes = createIncludesFromPolicyGenerateResult(
```

Comment:
Should we just do this once right after we create the extLocations on line 257?

Reply:
That would require another loop; we already have this loop here.

Reply:
Oh actually, I see the other loop earlier in the if block. I guess I could probably just pull that out.